README.md
# mock-servicer - 用于提供模拟数据的node服务器。 

<p>
  <a href="https://hai2007.gitee.io/npm-downloads?interval=7&packages=mock-servicer"><img src="https://img.shields.io/npm/dm/mock-servicer.svg" alt="downloads"></a>
  <a href="https://packagephobia.now.sh/result?p=mock-servicer"><img src="https://packagephobia.now.sh/badge?p=mock-servicer" alt="install size"></a>
  <a href="https://www.npmjs.com/package/mock-servicer"><img src="https://img.shields.io/npm/v/mock-servicer.svg" alt="Version"></a>
  <a href="https://github.com/hai2007/mock-servicer/blob/master/LICENSE"><img src="https://img.shields.io/npm/l/mock-servicer.svg" alt="License"></a>
  <a href="https://github.com/hai2007/mock-servicer">
      <img alt="GitHub repo stars" src="https://img.shields.io/github/stars/hai2007/mock-servicer?style=social">
  </a>
</p>

## Issues
使用的时候遇到任何问题或有好的建议，请点击进入[issue](https://github.com/hai2007/mock-servicer/issues)！

> 基于此项目开发的[mock-manager](https://github.com/hai2007/mock-manager)是一个更简单的提供模拟数据的“新增”、“删除”、“更新”和“使用”于一体的项目，使用起来非常简单，推荐使用此项目哦~

## 如何使用？

```
npm install mock-servicer
```

安装好了以后，引入并启动即可：

```js
const MockServicer = require('mock-servicer');

MockServicer({

  // 请求端口，可选，默认8080
  port: 8080,

  // 服务器根地址，可选，默认当前路径
  contentBase:'./',

  // mock数据缓存根地址，可选，默认当前路径
  mockBase:'./mock',

  // 404的提示界面，可选
  template404:function(list){
    // 返回当前路径下文件列表
    return template;
  },

  // 自定义处理，可选
  handler:function(options){

    return {
      code:"",// 响应码，可选，默认200
      data:"",// 数据，必输
      type:""// 数据类型，可选，默认application/json
    };

  },

  // 请求拦截，可选，默认全部同意
  intercept:function(options){
      // 如果返回false，服务器会拒绝这次请求
      return true|false;
  }

});
```

### 和服务器交互

服务器启动成功以后，我们就可以通过请求的方式和数据服务器进行数据交互了。

交互主要分为下列几种：

- 新增或更新

也就是在仓库中新增或修改已经有的模拟数据

> 其中url用于指定模拟数据名称，method用于指定模拟数据请求方法，总之，url和method唯一确定了哪一方模拟数据（主要是考虑到实际项目的同一个url可能会使用不同的method实现不同的功能），下同。

```js
$.ajax({
  url: 'http://127.0.0.1:8080/update?url=XXX&method=XXX',
  type: "POST",
  data: "需要保存的数据"
});
```

- 删除

删除已经存在的模拟数据

```js
$.ajax({
  url: 'http://127.0.0.1:8080/delete?url=XXX&method=XXX'
});
```

- 查询

如果需要使用这份模拟数据，可以这样请求，返回的是mock运行后的结果而不是你新增的原始代码

```js
$.ajax({
  url: 'http://127.0.0.1:8080/query?url=XXX&method=XXX'
});
```

- Mock查询

和```查询```类似，只是这里直接使用```XXX```指定请求的模拟数据文件名称

```js
$.ajax({
  url: 'http://127.0.0.1:8080/mock?XXX'
});
```

- 原始查询

有时候，我们可能希望开发一个平台来管理我们的模拟数据，这个方法就可以返回原始代码，方便我们在线维护（比如这个项目：[mock-manager](https://github.com/hai2007/mock-manager)）

```js
$.ajax({
  url: 'http://127.0.0.1:8080/oralquery?url=XXX&method=XXX'
});
```

- 自定义处理

对于一些特殊情况，我们无法考虑那么多，比如访问权限等，这时候就可以自定义处理规则了

```js
$.ajax({
  url: 'http://127.0.0.1:8080/handler'
});
```

当前，这意味着启动服务器时```handler```配置项是必须的。

上面的```需要保存的数据```可以是一个普通的JSON字符串，比如：

```js
data: `{
  "key1": "value1",
  "key2": "value2",
  "key3": "value3"
}`
```

或者使用mock，比如：

```js
data: `Mock.mock({
// 属性 list 的值是一个数组，其中含有 1 到 10 个元素
  'list|1-10': [{
    // 属性 id 是一个自增数，起始值为 1，每次增 1
    'id|+1': 1
  }]
})`
```

### 作为普通的数据服务器

除了上面特殊情况的交互外，你还可以直接访问位于```服务器根地址 contentBase```下的文件。

## 命令行启动

除了作为一个node包使用外，还可以在安装以后使用命令行启动：

```bash
mock-servicer --config ./mock-servicer.config.js
```

需要指定一个配置文件用来规定如何启动，配置文件格式如下：

```js
// mock-servicer.config.js
module.exports = {

  port: 8080,
  // 和直接作为包使用的时候可配置参数的一致的

};
```

开源协议
---------------------------------------
[MIT](https://github.com/hai2007/mock-servicer/blob/master/LICENSE)

Copyright (c) 2021-2022 [hai2007](https://hai2007.gitee.io/sweethome/) 走一步，再走一步。


CHANGELOG
v0.1.0:
    date:2021-05-06
    changes:
      -  初始化版本
v0.1.1:
    date:2021-05-06
    changes:
      -  添加跨域解决方案
v1.0.0:
    date:2021-05-07
    changes:
      -  对项目进行一次大的优化升级（非兼容改造）
        1.新增和服务器交互方法
        2.可以作为普通的数据服务器使用
v1.0.1:
    date:2021-05-07
    changes:
      -  修复服务器内部意外错误导致停机问题
v1.1.0:
    date:2021-05-07
    changes:
      -  添加404的模板配置
v1.2.0:
    date:2021-05-08
    changes:
      -  功能加强 
        1.自定义处理把柄handler
        2.新增请求是否拒绝拦截把柄
      -  新增命令行操作
v1.2.2:
    date:2021-05-10
    changes:
      -  新增查询原始数据的方法 ```oralquery```
v1.2.3:
    date:2021-05-10
    changes:
      -  修复无成功回调的问题
v1.2.4:
    date:2021-06-23
    changes:
      -  修复数据修改后查询返回缓存数据错误
v1.3.0:
    date:2022-03-03
    changes:
      -  新增底层依赖 ```@hai2007/nodejs``` 版本（此依赖有部分非完全兼容，但本项目的升级保证向下兼容，特此备注）
      -  新增mock查询的方法 ```mock```

test/index.js
const MockServicer = require('../index');

MockServicer({

  // 请求端口，默认8080
  port: 30000,

  contentBase: './',

  mockBase: './mock',

  // template404:function(list){
  //   console.log(list);
  //   return "404";
  // }，

  handler: function (options) {

    // throw new Error('一个错误handler');

    return {
      code: "200",
      data: "{info:'handler的数据'}",
      // type:""
    };

  },

  // intercept:function(options){
  //   // throw new Error('一个错误intercept');
  //     // 如果返回false，服务器会拒绝这次请求
  //     return true;
  // }

});

test/mock-servicer.config.js
module.exports = {

  port: 20000,

};

test/mock/mock-GET-78d6b5ea.js
module.exports=function(Mock){return Mock.mock({
      // 属性 list 的值是一个数组，其中含有 1 到 10 个元素
        'list|1-10': [{
          // 属性 id 是一个自增数，起始值为 1，每次增 1
          'id|+1': 1
        }]
      });};

mime.types.js
module.exports = {
  "html": "text/html",
  "htm": "text/html",
  "shtml": "text/html",
  "css": "text/css",
  "xml": "text/xml",
  "gif": "image/gif",
  "jpeg": "image/jpeg",
  "jpg": "image/jpeg",
  "js": "application/javascript",
  "atom": "application/atom+xml",
  "rss": "application/rss+xml",

  "mml": "text/mathml",
  "txt": "text/plain",
  "jad": "text/vnd.sun.j2me.app-descriptor",
  "wml": "text/vnd.wap.wml",
  "htc": "text/x-component",

  "png": "image/png",
  "tif": "image/tiff",
  "tiff": "image/tiff",
  "wbmp": "image/vnd.wap.wbmp",
  "ico": "image/x-icon",
  "jng": "image/x-jng",
  "bmp": "image/x-ms-bmp",
  "svg": "image/svg+xml",
  "svgz": "image/svg+xml",
  "webp": "image/webp",

  "woff": "application/font-woff",
  "jar": "application/java-archive",
  "war": "application/java-archive",
  "ear": "application/java-archive",
  "json": "application/json",
  "hqx": "application/mac-binhex40",
  "doc": "application/msword",
  "pdf": "application/pdf",
  "ps": "application/postscript",
  "eps": "application/postscript",
  "ai": "application/postscript",
  "rtf": "application/rtf",
  "m3u8": "application/vnd.apple.mpegurl",
  "xls": "application/vnd.ms-excel",
  "eot": "application/vnd.ms-fontobject",
  "ppt": "application/vnd.ms-powerpoint",
  "wmlc": "application/vnd.wap.wmlc",
  "kml": "application/vnd.google-earth.kml+xml",
  "kmz": "application/vnd.google-earth.kmz",
  "7z": "application/x-7z-compressed",
  "cco": "application/x-cocoa",
  "jardiff": "application/x-java-archive-diff",
  "jnlp": "application/x-java-jnlp-file",
  "run": "application/x-makeself",
  "pl": "application/x-perl",
  "pm": "application/x-perl",
  "prc": "application/x-pilot",
  "pdb": "application/x-pilot",
  "rar": "application/x-rar-compressed",
  "rpm": "application/x-redhat-package-manager",
  "sea": "application/x-sea",
  "swf": "application/x-shockwave-flash",
  "sit": "application/x-stuffit",
  "tcl": "application/x-tcl",
  "tk": "application/x-tcl",
  "der": "application/x-x509-ca-cert",
  "pem": "application/x-x509-ca-cert",
  "crt": "application/x-x509-ca-cert",
  "xpi": "application/x-xpinstall",
  "xhtml": "application/xhtml+xml",
  "xspf": "application/xspf+xml",
  "zip": "application/zip",

  "bin": "application/octet-stream",
  "exe": "application/octet-stream",
  "dll": "application/octet-stream",
  "deb": "application/octet-stream",
  "dmg": "application/octet-stream",
  "iso": "application/octet-stream",
  "img": "application/octet-stream",
  "msi": "application/octet-stream",
  "msp": "application/octet-stream",
  "msm": "application/octet-stream",

  "docx": "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
  "xlsx": "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
  "pptx": "application/vnd.openxmlformats-officedocument.presentationml.presentation",

  "mid": "audio/midi",
  "midi": "audio/midi",
  "kar": "audio/midi",
  "mp3": "audio/mpeg",
  "ogg": "audio/ogg",
  "m4a": "audio/x-m4a",
  "ra": "audio/x-realaudio",

  "3gpp": "video/3gpp",
  "3gp": "video/3gpp",
  "ts": "video/mp2t",
  "mp4": "video/mp4",
  "mpeg": "video/mpeg",
  "mpg": "video/mpeg",
  "mov": "video/quicktime",
  "webm": "video/webm",
  "flv": "video/x-flv",
  "m4v": "video/x-m4v",
  "mng": "video/x-mng",
  "asx": "video/x-ms-asf",
  "asf": "video/x-ms-asf",
  "wmv": "video/x-ms-wmv",
  "avi": "video/x-msvideo",
};

index.js
/*!
 * MOCK数据服务器
 * https://github.com/hai2007/mock-servicer
 *
 * author hai2007 < https://hai2007.gitee.io/sweethome >
 *
 * Copyright (c) 2021-2022 hai2007 走一步，再走一步。
 * Released under the MIT license
 */

const http = require('http');
const { log, fullPath } = require('@hai2007/nodejs');
const fs = require('fs');
const path = require('path');
const url = require('url');
const mineTypes = require('./mime.types.js');
const responseFileList = require('./tool/responseFileList.js');
const urlToString = require('./tool/urlToString');

const jsonfile = JSON.parse(fs.readFileSync(path.join(__dirname, './package.json')));

module.exports = function (config) {

  const port = 'port' in config ? config.port : 8080; // 端口号
  const basePath = fullPath(config.contentBase || "./", process.cwd());// 根路径
  const mockBasePath = fullPath(config.mockBase || "./", process.cwd());// 根路径

  http.createServer(function (request, response) {

    let contentType = 'application/json';
    let responseCode = '200';
    let responseData = "{}";

    let urlObject = url.parse(request.url);

    // 获取data
    require('./tool/getData')(request, data => {

      let options = {

        // 请求方法
        method: request.method,

        // url
        url: urlObject.pathname.replace(/^\//, '').replace(/\/$/, ''),

        // 请求参数
        query: require('./tool/toQuery.js')(urlObject.query),

        // 数据
        value: data

      };

      try {
        if ('intercept' in config && !config.intercept(options)) {
          responseCode = "501";
          responseData = "The current request is not supported.";
          contentType = 'text/plain';
        } else {

          // 用于区分这次请求的目的
          let preUrl = options.url.split('/')[0];

          // 新增或更新
          // POST localhost:8080/update?url=XXX&method=XXX
          if (preUrl == 'update') {

            let datapath = fullPath("./mock-" + urlToString(options.query.url, options.query.method) + ".js", mockBasePath);

            // 写入内容
            fs.writeFileSync(datapath, `module.exports=function(Mock){return ${options.value};};`, {
              encoding: 'utf8'
            });

          }

          // 删除
          // localhost:8080/delete?url=XXX&method=XXX
          else if (preUrl == 'delete') {

            let datapath = fullPath("./mock-" + urlToString(options.query.url, options.query.method) + ".js", mockBasePath);

            // 删除内容
            if (fs.existsSync(datapath)) {
              fs.unlinkSync(datapath);
            }

          }

          // 查询
          // localhost:8080/query?url=XXX&method=XXX
          else if (preUrl == 'query') {

            let datapath = fullPath("./mock-" + urlToString(options.query.url, options.query.method) + ".js", mockBasePath);

            if (fs.existsSync(datapath)) {
              responseData = JSON.stringify(require(datapath)(require('mockjs')));

              // 删除缓存
              delete require.cache[require.resolve(datapath)];
            } else {
              responseCode = "404";
            }

          }

          // Mock查询
          // localhost:8080/mock?XXX
          else if (preUrl == 'mock') {

            let datapath = fullPath("./" + urlObject.query + ".js", mockBasePath);

            if (fs.existsSync(datapath)) {
              responseData = JSON.stringify(require(datapath)(require('mockjs')));

              // 删除缓存
              delete require.cache[require.resolve(datapath)];
            } else {
              responseCode = "404";
            }

          }

          // 原始查询
          // localhost:8080/oralquery?url=XXX&method=XXX
          else if (preUrl == 'oralquery') {

            let datapath = fullPath("./mock-" + urlToString(options.query.url, options.query.method) + ".js", mockBasePath);

            if (fs.existsSync(datapath)) {
              responseData = (fs.readFileSync(datapath, 'utf-8') + "").replace(/^module\.exports=function\(Mock\)\{return /, '').replace(/;\};$/, '');
              contentType = 'text/plain';
            } else {
              responseCode = "404";
            }

          }

          // 自定义处理
          // localhost:8080/handler
          else if (preUrl == 'handler') {

            let resultData = config.handler(options);

            if ('code' in resultData) responseCode = resultData.code;
            responseData = resultData.data;
            if ('type' in resultData) contentType = resultData.type;

          }

          // 默认就作为普通的数据服务器
          else {

            // 请求的文件路径
            let filePath = fullPath(options.url == "" ? "index.html" : options.url, basePath);

            // 文件后缀名称
            let dotName = /\./.test(filePath) ? filePath.match(/\.([^.]+)$/)[1] : "";

            // 文件类型
            if (dotName != "") contentType = mineTypes[dotName];

            // 如果需要读取的文件存在
            if (fs.existsSync(filePath) && !fs.lstatSync(filePath).isDirectory()) {
              responseData = fs.readFileSync(filePath);
            }

            // 如果不存在，就返回404并列举出当前目录内容
            // 这样的目的是为了方便二次开发
            else {
              responseCode = "404";
              responseData = ('template404' in config ? config.template404 : require('./tool/template404'))(responseFileList(filePath));
              contentType = "text/html";
            }

          }

        }
      }

      // 可能出错
      catch (e) {
        responseCode = "500";
        responseData = "" + e;
        contentType = 'text/plain';
      }

      response.writeHead(responseCode, {

        // 设置跨域
        "Access-Control-Allow-Origin": "*",// 允许的域
        "Access-Control-Allow-Headers": "*",// 允许的header类型
        "Access-Control-Allow-Methods": "*",// 允许的请求方法

        // 标记服务器名称
        "X-Powered-By": jsonfile.name + " " + jsonfile.version,

        // 响应内容类型
        "Content-Type": contentType + ";charset=utf-8"

      });

      response.write(responseData);

      response.end();

    });

  })

    // 启动监听
    .listen(port);

  log(jsonfile.name + ' running on port:' + port);

};